### ISSUE TITLE
Failing with error Could not load file or assembly 'Microsoft.TeamFoundation.DistributedTask.Agent.Interfaces'

### ISSUE THREAD

**Issue by ghost** [XpiritBV/Xpirit-Vsts-Release-Settings#2]
The task started failing last night, no changes had been made to any build or release definiitons.

Full Call Stack:
2016-08-25T00:03:53.2157766Z ##[error]System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.TeamFoundation.DistributedTask.Agent.Interfaces, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
     2016-08-25T00:03:53.2167768Z File name: 'Microsoft.TeamFoundation.DistributedTask.Agent.Interfaces, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
     2016-08-25T00:03:53.2167768Z    at System.Signature.GetSignature(Void\* pCorSig, Int32 cCorSig, RuntimeFieldHandleInternal fieldHandle, IRuntimeMethodInfo methodHandle, RuntimeType declaringType)
     2016-08-25T00:03:53.2167768Z    at System.Reflection.RuntimePropertyInfo.get_Signature()
     2016-08-25T00:03:53.2167768Z    at System.RuntimeType.RuntimeTypeCache.MemberInfoCache`1.PopulateProperties(Filter filter, RuntimeType declaringType, Dictionary`2 csPropertyInfos, Boolean[] usedSlots, ListBuilder`1& list)
     2016-08-25T00:03:53.2167768Z    at System.RuntimeType.RuntimeTypeCache.MemberInfoCache`1.PopulateProperties(Filter filter)
     2016-08-25T00:03:53.2167768Z    at System.RuntimeType.RuntimeTypeCache.MemberInfoCache`1.GetListByName(Char* pName, Int32 cNameLen, Byte* pUtf8Name, Int32 cUtf8Name, MemberListType listType, CacheType cacheType)
     2016-08-25T00:03:53.2167768Z    at System.RuntimeType.RuntimeTypeCache.MemberInfoCache`1.Populate(String name, MemberListType listType, CacheType cacheType)
     2016-08-25T00:03:53.2167768Z    at System.RuntimeType.GetPropertyCandidates(String name, BindingFlags bindingAttr, Type[] types, Boolean allowPrefixLookup)
     2016-08-25T00:03:53.2167768Z    at System.RuntimeType.GetProperties(BindingFlags bindingAttr)
     2016-08-25T00:03:53.2167768Z    at System.Management.Automation.InternalParameterMetadata.ConstructCompiledParametersUsingReflection(Boolean processingDynamicParameters)
     2016-08-25T00:03:53.2167768Z    at System.Management.Automation.InternalParameterMetadata.Get(Type type, ExecutionContext context, Boolean processingDynamicParameters)
     2016-08-25T00:03:53.2167768Z    at System.Management.Automation.CommandMetadata..ctor(String commandName, Type cmdletType, ExecutionContext context)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.CommandMetadata.Get(String commandName, Type cmdletType, ExecutionContext context)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.CmdletInfo.get_CommandMetadata()
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.CommandProcessor.NewParameterBinderController(InternalCommand command)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.CommandProcessor.get_CmdletParameterBinderController()
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.CommandProcessor.Prepare(IDictionary psDefaultParameterValues)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.CommandProcessorBase.DoPrepare(IDictionary psDefaultParameterValues)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.Internal.PipelineProcessor.Start(Boolean incomingStream)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input, Hashtable errorResults, Boolean enumerate)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.PipelineOps.InvokePipeline(Object input, Boolean ignoreInput, CommandParameterInternal[][] pipeElements, CommandBaseAst[] pipeElementAsts, CommandRedirection[][] commandRedirections, FunctionContext funcContext)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.Interpreter.ActionCallInstruction`6.Run(InterpretedFrame frame)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
     2016-08-25T00:03:53.2177767Z    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
     2016-08-25T00:03:53.2177767Z
     2016-08-25T00:03:53.2177767Z WRN: Assembly binding logging is turned OFF.
     2016-08-25T00:03:53.2177767Z To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog](DWORD) to 1.
     2016-08-25T00:03:53.2187766Z Note: There is some performance penalty associated with assembly bind failure logging.
     2016-08-25T00:03:53.2187766Z To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
     2016-08-25T00:03:53.2187766Z
     2016-08-25T00:04:03.9083379Z
          2016-08-25T00:04:03.9270768Z ##[error]PowerShell script completed with 1 errors

**Comment by jessehouwing** [comment]
Can you check whether you agent performed an auto-update? In that case
you'd have an agent.old folder.

**Comment by ghost** [comment]
I am using VSTS and hosted build agents, where would I look to find out what version is being used?

**Comment by jessehouwing** [comment]
In the agent capabilities it on the agent queue or agent pool screen.

I think they may have switched to the new dotnet core based agent.

**Comment by ghost** [comment]
Agent version is : 2.105.2

**Comment by jessehouwing** [comment]
Yeah, that would require this task to be updated to the new powershell3
handler which can't read all secrets yet due to api  limitations.

I don't think you can currently force a lower agent version on the hosted
pool.

**Comment by ghost** [comment]
OK, are you planning on updating for powershell3 when you can work round the api limitations?

**Comment by jessehouwing** [comment]
It looks like the SDK has been updated with a `Get-VstsTaskVariableInfo` cmdlet, which should provide the required functionality. I just had a short conversation with PascalN and he'll be looking into this. No timelines until we have been able to perform some tests.

https://github.com/Microsoft/vsts-task-lib/blob/master/powershell/VstsTaskSdk/InputFunctions.ps1

**Comment by ghost** [comment]
Thanks for the very quick response, we have found this component very useful in keeping environment configuration as part of our releases and not azure operations managment.

I look forward to seeing the updated version when you have completed the testing.

**Comment by pascalnaber** [comment]
Status update:
I've created a branch with an upgrade which supports the new buildagent. This looks good on my local machine. Unfortunately I'm not able to upload the task because of some weird error messages about the supported buildagent. Hopefully a new release tomorrow. Jesse, thanks for the help!

**Comment by pascalnaber** [comment]
I've uploaded a new version that supports the new buildagent. The exception I was getting is a known issue and Microsoft will fix it next week. It's not possible to upload a task which supports the new buildagent. This results in the decision to supports buildagent 1.102 in the configuration of the task, so it could be uploaded. The code temporary checks for a buildagent version 2 or higher when running. 

The change to the new buildagent also resulted in strange behavior of script scope variables. Which is the reason I had to remove all functions. Hopefully I can refactor this soon.
